### PR TITLE
Background Cleaner: add scanning hash table background.

### LINKED ISSUE #197: [NEW] KVDK need to have the KV expire capability to support a lot of related Redis commands.
**The problem/use-case that the feature addresses**

One of Redis's key features is that it allows data structures to set an expire times. Currently KVDK do not provide the KV expire capability

**Description of the feature**

We expect KVDK to provide the option to set the expiry time when executing commands such as Set, Get, etc. This feature will also be used to implement the redis command such as command TTL, EXPIRE, etc. 

**Alternatives you've considered**

No

**Additional information**

No

### PR DESCRIPTION
**Pull request by sherry-1001**
### What problem does this PR solve?

Issue Number: close #197
Problem Summary:

### What is changed and how it works?
1. add scanning hash table, it is in `backgoundOldCleaner` background thread.
2. Add hash entry status to avoid repeatedly pushing old record into cleaner.
3. Replace `search` by `lookupKey`.
4. Add persist API.

Now  we can support expire function for string type. 
Notice: For expired collection, the foreground thread is not responsible to delete now. 

Tests 

- Unit test
    background cleaner test case for string type.
- Integration test
- No code

Side effects

- Performance regression
    - Consumes more CPU
    - Consumes more MEM
    - Consumes more PMEM

### CODE REVIEW

File: engine/c.cpp [hunk 1]
```diff
@@ -412,4 +412,8 @@ KVDKStatus KVDKGetTTL(KVDKEngine* engine, const char* str, size_t str_len,
                       int64_t* ttl_time) {
   return engine->rep->GetTTL(std::string(str, str_len), ttl_time);
 }
+
+KVDKStatus KVDKPersist(KVDKEngine* engine, const char* str, size_t str_len) {
```
Comment (JiayuZzz) [inline engine/c.cpp:416]: Will Persist be obscured by PMem Persist? Maybe just call it RemoveExpire or something?

File: tests/tests.cpp [hunk 2]
```diff
@@ -1863,7 +1864,7 @@ TEST_F(EngineBasicTest, TestExpireAPI) {
 
   std::string got_val;
   int64_t ttl_time;
-  WriteOptions write_options1{0 /* invalid argument */, false};
+  WriteOptions write_options1{1, false};
```
Comment (ZiyanShi) [inline tests/tests.cpp:1867]: 1 for what?

File: tests/tests.cpp [hunk 3]
```diff
@@ -1969,6 +1971,76 @@ TEST_F(EngineBasicTest, TestExpireAPI) {
...
+      std::string key = std::to_string(i) + "stringk";
+      std::string val = std::to_string(i) + std::to_string(id) + "stringval";
+      std::string got_val;
+      ASSERT_EQ(engine->Set(key, val, WriteOptions{INT32_MAX, false}),
+                Status::Ok);
+    }
+  };
+  auto ExpiredClean = [&](uint64_t id) {
+    auto test_kvengine = static_cast<KVEngine*>(engine);
+    test_kvengine->ExpiredCleaner();
```
Comment (ZiyanShi) [inline tests/tests.cpp:1991]: Don't name a function like an object.
CleanExpired

File: tests/tests.cpp [hunk 4]
```diff
@@ -1969,6 +1971,76 @@ TEST_F(EngineBasicTest, TestExpireAPI) {
...
+TEST_F(EngineBasicTest, TestBackGroundCleaner) {
+  configs.max_access_threads = 16;
+  configs.background_work_interval = 1000;
+  ASSERT_EQ(Engine::Open(db_path.c_str(), &engine, configs, stdout),
+            Status::Ok);
+  int cnt = 100;
+  auto SetString = [&](uint64_t id) {
+    for (int i = 0; i < cnt; ++i) {
+      std::string key = std::to_string(i) + "stringk";
+      std::string val = std::to_string(i) + std::to_string(id) + "stringval";
```
Comment (ZiyanShi) [inline tests/tests.cpp:1983]: Does this id really matter?

File: engine/utils/utils.hpp [hunk 5]
```diff
@@ -465,12 +466,30 @@ inline UnixTimeType unix_time(void) {
...
+  return true;
+}
+
+inline ExpireTimeType TTLToExpireTime(TTLType ttl_time,
+                                      int64_t base_time = millisecond_time()) {
+  return ttl_time == kPersistTime ? kPersistTime : ttl_time + base_time;
+}
+
+inline TTLType ExpireTimeToTTL(ExpireTimeType expire_time,
+                               int64_t base_time = millisecond_time()) {
```
Comment (ZiyanShi) [inline engine/utils/utils.hpp:490]: ExpireTime/UnixTime base_time

File: engine/kv_engine.cpp [hunk 6]
```diff
@@ -1344,123 +1353,133 @@ Status KVEngine::SGet(const StringView collection, const StringView user_key,
...
-    return s;
+  LookupResult res = lookupKey(str, ExpirableRecordType);
+  if (res.s == Status::Expired) {
+    // Push the expired record into cleaner and update hash entry status with
+    // HashEntryStatus::Expired.
+    // TODO(zhichen): This `if` will be removed when completing collection
+    // deletion.
+    if (res.entry_ptr->GetIndexType() == HashIndexType::StringRecord) {
+      hash_table_->UpdateEntryStatus(res.entry_ptr, HashEntryStatus::Expired);
+      delayFree(OldDeleteRecord{res.entry_ptr->GetIndex().ptr,
```
Comment (ZiyanShi) [inline engine/kv_engine.cpp:1367]: Don't delayFree here. If GetTTL() delayFree() an expired record, other API also should do this.

File: engine/kv_engine.cpp [hunk 7]
```diff
@@ -2009,9 +2041,8 @@ Snapshot* KVEngine::GetSnapshot(bool make_checkpoint) {
 }
 
 void KVEngine::delayFree(const OldDataRecord& old_data_record) {
-  old_records_cleaner_.Push(old_data_record);
-  // To avoid too many cached old records pending clean, we try to clean
-  // cached records while pushing new one
+  // To avoid too many cached old records pending clean, we try to clean cached
```
Comment (ZiyanShi) [inline engine/kv_engine.cpp:2044]: Where does my old_data_record go?

**Review by ZiyanShi — CHANGES_REQUESTED** [review]

File: tests/tests.cpp (hunk 2 above)
Comment (sherry-1001) [reply tests/tests.cpp:1867]: ttl_time is 1s. 0 is invalid, similar to redis.

File: engine/c.cpp (hunk 1 above)
Comment (sherry-1001) [reply engine/c.cpp]: removed.

File: tests/tests.cpp (hunk 3 above)
Comment (sherry-1001) [reply tests/tests.cpp:1991]: ok, rename.

File: engine/kv_engine.cpp (hunk 7 above)
Comment (sherry-1001) [reply engine/kv_engine.cpp:2036]: recovery.

File: tests/tests.cpp [hunk 8]
```diff
@@ -1969,6 +1971,76 @@ TEST_F(EngineBasicTest, TestExpireAPI) {
...
+  };
+  auto ExpiredClean = [&](uint64_t id) {
+    auto test_kvengine = static_cast<KVEngine*>(engine);
+    test_kvengine->CleanExpire();
+  };
+
+  auto ExpireString = [&](uint64_t id, Status s) {
+    for (int i = 0; i < cnt; ++i) {
+      std::string key = std::to_string(i) + "stringk";
+      std::string val = std::to_string(i) + std::to_string(id) + "stringval";
```
Comment (ZiyanShi) [inline tests/tests.cpp:1997]: What's the id and val for?

**Review by ZiyanShi — CHANGES_REQUESTED** [review]

**Review by ZiyanShi — APPROVED** [review]

File: tests/tests.cpp [hunk 9]
```diff
@@ -2034,7 +2031,7 @@ TEST_F(EngineBasicTest, TestBackGroundCleaner) {
   {
     std::vector<std::thread> ts;
     ts.emplace_back(std::thread(SetString, 0));
```
Comment (ZiyanShi) [inline tests/tests.cpp:2033]: If we don't really need those ids, don't add them. Other reviewers may expect those ids fill a special purpose, such as to establish dependency between threads, but they don't.
Please keep things as simple as possible.

**Review by ZiyanShi — APPROVED** [review]

File: engine/hash_table.hpp [hunk 10]
```diff
@@ -48,26 +55,38 @@ struct alignas(16) HashEntry {
...
+
   // Check if "key" of data type "target_type" is indexed by "this". If
   // matches, copy data entry of data record of "key" to "data_entry_metadata"
   // and return true, otherwise return false.
   bool Match(const StringView& key, uint32_t hash_k_prefix,
              uint16_t target_type, DataEntry* data_entry_metadata);
 
+ private:
+  // Make this hash entry empty while its content been deleted
+  void Clear() { header_.index_type = PointerType::Empty; }
```
Comment (JiayuZzz) [inline engine/hash_table.hpp:84]: clear()

File: engine/hash_table.hpp [hunk 11]
```diff
@@ -48,26 +55,38 @@ struct alignas(16) HashEntry {
...
+
+  bool UpdateEntryStatus(HashEntryStatus entry_status) {
```
Comment (JiayuZzz) [inline engine/hash_table.hpp:86]: updateEntryStatus

File: engine/hash_table.hpp [hunk 12]
```diff
@@ -138,14 +157,20 @@ class HashTable {
   //
   // entry_ptr: position to insert, it's get from SearchForWrite()
   void Insert(const KeyHashHint& hint, HashEntry* entry_ptr, RecordType type,
```
Comment (JiayuZzz) [inline engine/hash_table.hpp:159]: we should avoid default para in future

File: engine/kv_engine.cpp [hunk 13]
```diff
@@ -2111,6 +2123,50 @@ void KVEngine::backgroundDramCleaner() {
...
+    uint64_t need_clean_num = 0;
+    while (bucket_iter != end_bucket_iter) {
+      switch (bucket_iter->GetIndexType()) {
+        case PointerType::StringRecord: {
+          if (bucket_iter->IsTTLStatus() &&
+              bucket_iter->GetIndex().string_record->HasExpired()) {
+            hash_table_->UpdateEntryStatus(&(*bucket_iter),
+                                           HashEntryStatus::Expired);
+            // push expired cleaner
+            expired_record_queue.push_back(OldDeleteRecord{
```
Comment (JiayuZzz) [inline engine/kv_engine.cpp:2146]: the pmem record of OldDeleteRecord here is still data record type right?

**Review by JiayuZzz — APPROVED** [review]
LGTM

**Pull request merged**